drivers/wireless/lpwan/rn2xx3/rn2xx3.c

```diff
@@ -0,0 +1,1508 @@
+ *
+ * Description:
+ *    Get the modulation type of the RN2xx3.
+ *
+ ****************************************************************************/
+
+static int rn2xx3_getmod(FAR struct rn2xx3_dev_s *priv, FAR char *mod)
+{
+  int err;
+  char response[12];
+
+  err = file_write(&priv->uart, "radio get mod\r\n",
+                   sizeof("radio get mod\r\n") - 1);
+  if (err < 0)
+    {
+      return err;
+    }
+
+  err = read_line(priv, response, sizeof(response));
+
+  if (err < 0)
+    {
+      return err;
+    }
+
+  /* Copy everything except \r\n */
+
+  memcpy(mod, response, err - 2);
+  response[err - 2] = '\0';
+  return 0;
+}
+
+/****************************************************************************
+ * Name: rn2xx3_read
+ *
+ * Description:
+ *   Puts the RN2xx3 into continuous receive mode and waits for data to be
+ *   received before returning it.
+ *
+ ****************************************************************************/
+
+static ssize_t rn2xx3_read(FAR struct file *filep, FAR char *buffer,
+                           size_t buflen)
+{
+  FAR struct inode *inode = filep->f_inode;
+  FAR struct rn2xx3_dev_s *priv = inode->i_private;
+  ssize_t length = 0;
+  ssize_t received = 0;
+  static const char receive_cmd[] = "radio rx 0\r\n";
+  char response[30];
+  char hex_byte[3] =
+  {
+    0, 0 /* Extra space for null terminator */
+  };
+
+  /* Exclusive access */
+
+  length = nxmutex_lock(&priv->devlock);
+  if (length < 0)
+    {
+      return length;
+    }
+
+  /* If we're still marked as receiving and the file position in non-zero, we
+   * can skip initiating receive mode and just continue parsing the packet.
+   *
+   * If the file position is non-zero and we're not receiving, return 0 to
+   * indicate the end of the last packet.
+   *
+   * Otherwise, we're not receiving and haven't been receiving, so we
+   * initiate a new receive mode.
+   */
+
+  if (filep->f_pos > 0 && !priv->receiving)
+    {
+      filep->f_pos = 0;
+      return 0;
+    }
+  else if (filep->f_pos > 0 && priv->receiving)
+    {
+      goto parse_packet;
+    }
+
+  priv->receiving = true; /* About to receive */
+
+  /* Flush data */
+
+  length = radio_flush(priv);
+  if (length < 0)
+    {
+      wlerr("Could not flush RN2xx3\n");
+    }
+
+  /* Pause the MAC layer */
+
+  length = mac_pause(priv);
+  if (length < 0)
+    {
+      wlerr("Couldn't pause MAC layer\n");
+      goto early_ret;
+    }
+
+  /* Put the radio into indefinite receive mode */
+
+  length = file_write(&priv->uart, receive_cmd, sizeof(receive_cmd) - 1);
+  if (length < 0)
+    {
+      goto early_ret;
+    }
+
+  /* Check for okay receipt of command */
+
+  length = get_command_err(priv);
+  if (length < 0)
+    {
+      goto early_ret;
+    }
+
+  /* Begin slowly parsing the response to check if we either received data or
+   * had an error (timeout)
+   * 9 is the size of 'radio_rx ' or 'radio_err'.
+   */
+
+  for (uint8_t i = 0; i < 10; i++)
```

Comment:
Shouldn't user handle operation retries? Maybe caller does not want to wait for 10 consecutive timeouts to abort?

Reply:
Oh wait, I missread it. Still a question - why read one byte at a time?

Reply:
You're right, not necessary.